lib/web_ui/lib/src/engine/text/paragraph.dart

```diff
@@ -1354,6 +1370,37 @@ class EngineParagraphBuilder implements ui.ParagraphBuilder {
   }
 }
 
+/// Holds information for a placeholder in a paragraph.
+///
+/// [width], [height] and [baselineOffset] are expected to be already scaled.
+class ParagraphPlaceholder {
+  ParagraphPlaceholder(
+    this.width,
+    this.height,
+    this.alignment, {
+    required this.baselineOffset,
+    required this.baseline,
+  });
+
+  /// The scaled width of the placeholder.
+  final double width;
+
+  /// The scaled height of the placeholder.
+  final double height;
+
+  /// Specifies how the placeholder rectangle will be vertically aligned with
+  /// the surrounding text.
+  final ui.PlaceholderAlignment alignment;
+
+  /// When the [alignment] value is [ui.PlaceholderAlignment.baseline], the
+  /// [baselineOffset] indicates the distance from the baseline to the top of
+  /// the placeholder rectangle.
+  final double baselineOffset;
+
+  /// Dictates whether to use alphabetic or ideographic baseline.
+  final ui.TextBaseline? baseline;
```

Comment:
can you add comment on when baseline is null please (if the only item on line is widgetspan?).

Reply:
+1. Alternatively, you could leave a link to https://api.flutter.dev/flutter/dart-ui/ParagraphBuilder/addPlaceholder.html that points to the upstream docs for the semantics being implemented here.

Reply:
Done.